specs/derivation.md

```diff
@@ -27,181 +43,606 @@ TODO
-This ensures at most a span of `CHANNEL_TIMEOUT` seconds of L1 data is required to reconstruct any channel-bank state in time.
+(where `++` denotes concatenation)
 
-### L1 transaction data format:
+The `rollup_payload` may be right-padded with 0s, which will be ignored. It's allowed for them to be
+interpreted as frames for channel 0, which must always be ignored.
+
+> **TODO** specify batcher authentication (i.e. where do we store / make available the public keys of authorize batcher
+> signers)
+
+### Frame Format
+
+A [channel frame][g-channel-frame] is encoded as:
 
 ```text
-transaction_data = version_byte | rollup_payload
+frame = channel_id ++ frame_number ++ frame_data_length ++ frame_data ++ is_last
+
+channel_id        = random ++ timestamp
+random            = bytes32
+timestamp         = uvarint
+frame_number      = uvarint
+frame_data_length = uvarint
+frame_data        = bytes
+is_last           = bool
+```
 
-# version_byte == 0:
-rollup_payload = frame ...   #  one or more frames, concatenated
+where:
+
+- `uvarint` is a variable-length encoding of a 64-bit unsigned integer into between 1 and 9 bytes, [as specified in
+  SQLite 4][sqlite-uvarint].
+- `channel_id` uniquely identifies a channel as the concatenation of a random value and a timestamp
+    - `random` is a random value such that two channels with different batches should have a different random value
+    - `timestamp` is the time at which the channel was created (UNIX time in seconds)
+    - The ID includes both the hash and the timestamp, in order to prevent a malicious sequencer from reusing the random
+      value after the channel has [timed out][g-channel-timeout] (refer to the [batcher specification][batcher-spec] to
+      learn more about channel timeouts). This will also allow us substitute `random` by a hash commitment to the
+      batches, should we want to do so in the future.
+- `frame_number` identifies the index of the frame within the channel (in bytes)
+- `frame_data_length` is the length of `frame_data` in bytes
+- `frame_data` is a sequence of bytes belonging to the channel, at index `max(frame_number-1, 0)`
+- `is_last` is a single byte with a value of 1 if the frame is the last in the channel, 0 if there are frames in the
+  channel. Any other value makes the frame invalid (it must be ignored by the rollup node).
+
+> **TODO** why is frame_number not 0 based?
+
+> **TODO** channel timestamps must be validated to not be in the future - is that currently done? + specify
+> Also: is this even safe? i.e. could L1 block producers manipulate the timestamp?
+
+[sqlite-uvarint]: https://www.sqlite.org/src4/doc/trunk/www/varint.wiki
+[batcher-spec]: batching.md
+
+### Channel Format
+
+A channel is encoded as `channel_encoding`, defined as:
+
+```text
+rlp_batches = []
+for batch in batches:
+    rlp_batches.append(rlp_encode(batch))
+channel_encoding = compress(rlp_batches)
+```
```

Comment:
The above is not correct: `batch` has a version byte which is not rlp-encoded, only the payload is.